crates/blockchain/payload.rs

```diff
@@ -145,19 +144,327 @@ fn calc_excess_blob_gas(parent_excess_blob_gas: u64, parent_blob_gas_used: u64)
     }
 }
 
-/// Calculates the total fees paid by the payload block
-/// Only potential errors are storage errors which should be treated as internal errors by rpc providers
-pub fn payload_block_value(block: &Block, storage: &Store) -> Option<U256> {
-    let mut total_fee = U256::zero();
-    let mut last_cummulative_gas_used = 0;
-    for (index, tx) in block.body.transactions.iter().enumerate() {
-        // Execution already succeded by this point so we can asume the fee is valid
-        let fee = tx.effective_gas_tip(block.header.base_fee_per_gas)?;
-        let receipt = storage
-            .get_receipt(block.header.number, index as u64)
-            .ok()??;
-        total_fee += U256::from(fee) * (receipt.cumulative_gas_used - last_cummulative_gas_used);
-        last_cummulative_gas_used = receipt.cumulative_gas_used;
+pub struct PayloadBuildContext<'a> {
+    pub payload: &'a mut Block,
+    pub evm_state: &'a mut EvmState,
+    pub remaining_gas: u64,
+    pub blob_count: u64,
+    pub receipts: Vec<Receipt>,
+    pub block_value: U256,
+    base_fee_per_blob_gas: U256,
+}
+
+impl<'a> PayloadBuildContext<'a> {
+    fn new(payload: &'a mut Block, evm_state: &'a mut EvmState) -> Self {
+        PayloadBuildContext {
+            remaining_gas: payload.header.gas_limit,
+            blob_count: 0,
+            receipts: vec![],
+            block_value: U256::zero(),
+            base_fee_per_blob_gas: U256::from(calculate_base_fee_per_blob_gas(
+                payload.header.excess_blob_gas.unwrap_or_default(),
+            )),
+            payload,
+            evm_state,
+        }
+    }
+}
+
+impl<'a> PayloadBuildContext<'a> {
+    fn parent_number(&self) -> BlockNumber {
+        self.payload.header.number - 1
+    }
+
+    fn block_number(&self) -> BlockNumber {
+        self.payload.header.number
+    }
+
+    fn store(&self) -> &Store {
+        self.evm_state.database()
+    }
+
+    fn base_fee_per_gas(&self) -> Option<u64> {
+        self.payload.header.base_fee_per_gas
     }
-    Some(total_fee)
+}
+
+/// Completes the payload building process, return the block value
+pub fn build_payload(payload: &mut Block, store: &Store) -> Result<U256, ChainError> {
+    info!("Building payload");
```

Comment:
I think logging what happens during payload building can bring good insight on how the block building process works and be of good value when errors arise. If the level of output is too annoying we can use `debug` instead of `info`

Reply:
Yes, let's use debug instead.